resolve.c

```diff
@@ -62,68 +70,93 @@ char *resolveHostname(const char *addr)
 	return hostname;
 }
 
-// This routine is run *after* garbage cleaning (default interval is once per hour)
-// to account for possibly updated hostnames
-void reresolveHostnames(void)
+// Resolve client host names
+void resolveClients(bool onlynew)
 {
-	int clientID;
-	for(clientID = 0; clientID < counters.clients; clientID++)
+	int i;
+	for(i = 0; i < counters.clients; i++)
 	{
 		// Memory validation
-		validate_access("clients", clientID, true, __LINE__, __FUNCTION__, __FILE__);
-
-		// Process this client only if it has at least one active query in the log
-		if(clients[clientID].count < 1)
-			continue;
+		validate_access("clients", i, true, __LINE__, __FUNCTION__, __FILE__);
 
-		// Get client hostname
-		char *hostname = resolveHostname(clients[clientID].ip);
-		if(strlen(hostname) > 0)
+		// If onlynew flag is set, we will only resolve new clients
+		// If not, we will try to re-resolve all known clients
+		if(onlynew && clients[i].new)
 		{
-			// Delete possibly already existing hostname pointer before storing new data
-			if(clients[clientID].name != NULL)
-			{
-				free(clients[clientID].name);
-				clients[clientID].name = NULL;
-			}
-
-			// Store client hostname
-			clients[clientID].name = strdup(hostname);
-			clients[clientID].new = false;
+			char *hostname = resolveHostname(clients[i].ip);
+
+			enable_thread_lock();
+
+			if(clients[i].name != NULL)
+				free(clients[i].name);
+
+			clients[i].name = hostname;
+			clients[i].new = false;
+
+			disable_thread_lock();
 		}
-		free(hostname);
 	}
 }
 
-// This routine is run *before* saving to the database (default interval is once per minute)
-// to account for new clients (and forward destinations)
-void resolveNewClients(void)
+// Resolve upstream destination host names
+static void resolveForwardDestinations(bool onlynew)
```

Comment:
We have the capability to resolve all forward destinations instead of just new ones, but we don't use it currently. When all clients are re-resolved, we should also re-resolve forward destination names.

Reply:
I wonder if this is really meaningful, as we don't expect the forward destination host names to ever change, do we? In contrast, users may decide to rename their local devices, e.g., from `android-a479b97ef09880b890d233a` to `paul-iphone`

Reply:
Perhaps if the upstream is localhost, and the hostname changes? It may not be worth implementing, but on the other hand, it is a very small cost (usually only 1 to 3 upstreams).